drivers/segger/stream_rtt.c

```diff
@@ -0,0 +1,192 @@
+ * Name: rttstream_gets
+ ****************************************************************************/
+
+static int rttstream_gets(FAR struct lib_instream_s *this,
+                          FAR void * buffer, int size)
+{
+  FAR struct lib_rttinstream_s *stream =
+                                (FAR struct lib_rttinstream_s *)this;
+  int ret;
+
+  DEBUGASSERT(stream);
+  ret = SEGGER_RTT_Read(stream->channel, buffer, size);
+  stream->public.nget += ret;
+  return ret;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: lib_rttoutstream_open
+ *
+ * Description:
+ *   Initializes a stream for use with the configured RTT interface.
+ *
+ * Input Parameters:
+ *   stream - User allocated, uninitialized instance of struct
+ *            lib_rttoutstream_s to be initialized.
+ *   channel - SEGGER RTT channel number
+ *   bufsize - Size of the RTT buffer
+ *
+ * Returned Value:
+ *   None (User allocated instance initialized).
+ *
+ ****************************************************************************/
+
+void lib_rttoutstream_open(FAR struct lib_rttoutstream_s *stream,
+                           int channel, size_t bufsize)
+{
+  if (channel)
+    {
+      stream->buffer = (FAR char *)kmm_malloc(bufsize);
+      DEBUGASSERT(stream->buffer);
+      sprintf(stream->name, "rtt%d", channel);
+      SEGGER_RTT_ConfigUpBuffer(channel, stream->name, stream->buffer,
+                            bufsize, SEGGER_RTT_MODE_DEFAULT);
+    }
+
+  stream->public.put = rttstream_putc;
+  stream->public.puts = rttstream_puts;
+  stream->public.flush = lib_noflush;
+  stream->public.nput = 0;
+  stream->channel = channel;
+}
+
+/****************************************************************************
+ * Name: lib_rttoutstream_close
+ ****************************************************************************/
+
+void lib_rttoutstream_close(FAR struct lib_rttoutstream_s *stream)
+{
+  kmm_free(stream->buffer);
+  SEGGER_RTT_ConfigUpBuffer(stream->channel, NULL, NULL,
+                            0, SEGGER_RTT_MODE_DEFAULT);
+}
+
+/****************************************************************************
+ * Name: lib_rttinstream
+ *
+ * Description:
+ *   Initializes input stream for use with the configured RTT interface.
+ *
+ * Input Parameters:
+ *   stream - User allocated, uninitialized instance of struct
+ *            lib_rttinstream_s to be initialized.
+ *   channel - SEGGER RTT channel number
+ *   bufsize - Size of the RTT input buffer
+ *
+ ****************************************************************************/
+
+void lib_rttinstream_open(FAR struct lib_rttinstream_s *stream,
+                          int channel, size_t bufsize)
```

Comment:
should we use BUFFER_SIZE_DOWN?

Reply:
or add below before line 169:
bufsize = bufsize ? bufsize : SEGGER_BUFFER_SIZE_DOWN;